xsync/asyncinit.py

```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2022-present, Ethan Henderson
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import inspect
+import logging
+import typing as t
+
+_log = logging.getLogger(__name__)
+
+
+class AsyncInitMixin:
+    __async_init__: t.Callable[..., t.Any]
```

Comment:
This typehint also doesn't pass pyright, and i'm not entirely sure why. I was unable to find a work around for this yet.

Reply:
Mypy complained at anything else, and having a NotImplemented version isn't an option at it complains about signatures.

I spose we need to decide at this stage whether to use Pyright alongside Mypy, as if so, we need to include it in the sessions.

Reply:
Running Pyright myself, I don't get any errors here. Is there a specific setting or configuration you're using?

Reply:
This one isn't an error inside the class. it's an error the end user gets when they override that method in their own implementation. You know me, I use both everywhere.

Reply:
i can handle integrating pyright if you like.

Reply:
Oh I see. And yeah sure -- I'm about to push commits relating to the other two fixes, then perhaps we can merge and you can add Pyright from there? I feel that might be cleaner than you committing directly to this PR.

Reply:
I also think pyright support is outside the scope of this PR. Sounds good.

Reply:
Sounds good. I'll resolve this convo, then you can make sure the other parts are all okay, then I spose make a PR of your own with the new Pyright support. I'd say include this patch in that PR as well.